### PR TITLE
milestone/nogoods

### LINKED ISSUE #64: Representation of nogoods
**Is your feature request related to a problem? Please describe.**
Represent nogoods in an efficient way, so further conclusions can be drawn.

**Describe the solution you'd like**
An encoding with bitmaps looks promising.

**Describe alternatives you've considered**
Alternatively some bdd-based approaches might be possible. Does not look like a natural fit of techniques.

**Additional context**
This is needed to allow a global collection of no goods.

### LINKED ISSUE #73: Alternative NoGoodLearning with the newly created NoGood representations
**Is your feature request related to a problem? Please describe.**
An algorithm, which will learn even more than the current implementation of the search space exploitation by using a more robust and better memorable set of NoGoods. 

**Describe the solution you'd like**
Usage of NoGoods and the NoGoodStore #65 for a less recursive approach.

**Describe alternatives you've considered**
This is an alternative to the current implementation.

**Additional context**
We hope to further improve the performance of adf-bdd by adding this feature. In any case we will see if and how much the performance will change with this alternative approach.

### PR DESCRIPTION
**Pull request by ellmau**
# What does this PR do?

* Resolves #73 
* Closes #64 

Implementation of an alternative NoGoodStore to easier learn and imply conclusions.

Update to version 0.2.5

# Checklist before creating a non-draft PR

- [x] All tests are passing
- [x] Clippy has no complains
- [x] Code is `rustfmt` formatted
- [x] Applicable labels are chosen (Note: it is not necessary to replicate the labels from the related issues)
- [x] There are no other open [Pull Requests](https://github.com/ellmau/adf-obdd/pulls) for the same update/change.
  - [ ] If there is a good reason to have another PR for the same update/change, it is well justified.

# Checklist on Guidelines and Conventions

- [x] Commit messages follow our guidelines
- [x] Code is self-reviewed
- [x] Naming conventions are met
- [x] New features are tested
  - [ ] `quickcheck` has been considered
  - [ ] All variants are considered and checked
- Clippy Compiler-exceptions
  - [x] Used in a sparse manner
  - [x] If used, a separate comment describes and justifies its use
- [x] `rustdoc` comments are self-reviewed and descriptive
- Error handling
  - [x] Use of `panic!(...)` applications is justified on non-recoverable situations
  - [x] `expect(...)` is used over `unwrap()` (except obvious test-cases)
- [x] No unsafe code (exceptions need to be discussed specifically)

### CODE REVIEW

File: lib/src/nogoods.rs [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+            None
+        }
+    }
+
+    /// Creates an updated [Vec<Term>], based on the given [&[Term]] and the [NoGood].
+    /// The parameter _update_ is set to [`true`] if there has been an update and to [`false`] otherwise
+    pub fn update_term_vec(&self, term_vec: &[Term], update: &mut bool) -> Vec<Term> {
+        *update = false;
+        term_vec.iter().enumerate().map(|(idx,val)|{
+	    let idx:u32 = idx.try_into().expect("no-good learner implementation is based on the assumption that only u32::MAX-many variables are in place");
```
Comment (mmarx) [inline lib/src/nogoods.rs:89]: Break the string literal to work around rust-lang/rustfmt#3863, which prevents formatting of the remainder of the function.

```suggestion
                let idx: u32 = idx.try_into().expect(
                    "no-good learner implementation is based on the assumption \
                     that only u32::MAX-many variables are in place",
                );
```

**Review by ellmau — COMMENTED** [review]
LGTM - Nice refactor to avoid the box and allow a functionpointer to be responsible as the heuristics functions.

**Comment by ellmau** [comment]
### Next step:
the consistency of choices and learnt information should be verified with respect to the adf more often to avoid local hill optimisation effects

**Review by mmarx — CHANGES_REQUESTED** [review]
lmgtm, I just have a few remarx.

File: lib/src/nogoods.rs [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+}
+
+impl From<&[Term]> for NoGood {
+    fn from(term_vec: &[Term]) -> Self {
+        Self::from_term_vec(term_vec)
+    }
+}
+
+/// A structure to store [NoGoods][NoGood] and offer operations and deductions based on them.
+// TODO:make struct crate-private
```
[the remark this answers is not in the document]
Comment (ellmau) [reply lib/src/nogoods.rs]: Yes it is no longer a goal for this PR and so the comment got removed

File: lib/src/nogoods.rs [hunk 3]
```diff
@@ -0,0 +1,824 @@
...
+    }
+
+    /// Updates the [NoGood] and a second one in a disjunctive (bitor) manner.
+    pub fn disjunction(&mut self, other: &NoGood) {
+        self.active = self.active.borrow().bitor(&other.active);
+        self.value = self.value.borrow().bitor(&other.value);
+    }
+
+    /// Returns [true] if the other [NoGood] matches with all the assignments of the current [NoGood].
+    pub fn is_violating(&self, other: &NoGood) -> bool {
```
[the remark this answers is not in the document]
Comment (ellmau) [reply lib/src/nogoods.rs]: Fixed by b3b2632.

File: lib/src/nogoods.rs [hunk 4]
```diff
@@ -0,0 +1,824 @@
...
+            // if the state is not new and the value is changed
+            if !is_new && upd {
+                return None;
+            }
+        }
+        if visit {
+            Some(result)
+        } else {
+            None
+        }
```
[the remark this answers is not in the document]
Comment (ellmau) [reply lib/src/nogoods.rs:86]: Thanks for the hint - flake got bumped just now (fac10dd) and therefore 1.62 is the current edition.

**Review by mmarx — APPROVED** [review]

**Pull request merged**